### PR TITLE
32bit Arrow sizes support & conversion functions

### PR DESCRIPTION
**Pull request by Klaim**
Adds Arrow length <-> std size/offset conversions functions with tests

This change does not however uses these functions in the rest of the library.
This builds but can lead to size truncations in 32bit at runtime, which are not visible in current tests.

We introduce 2 configuration options to enable the user to optionally constrain arrow data size to signed 32bit values,
and optionally enable runtime checks of arrow data size values.
The checks only happen in the new conversion functions so will not be effective until used in the rest of the library.

### CODE REVIEW

**Comment by Klaim** [comment]
I'll continue the addition of the size/offset Arrow<->std conversion functions in here, although for now it is not completely testable without allocating massive arrow data. There might be a way to test this case with special allocators though.

**Review by JohanMabille — APPROVED** [review]

File: include/sparrow/types/arrow_length.hpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    /// Type representing an Arrow length (sizes, offsets) in the Arrow specification and storage.
+    /// This is used internally but is not always directly convertible to `std::size_t`
+    /// and `std::ptrdiff_t` which are the "native" types to represent sizes and offsets.
+    /// For conversion purposes:
+    /// - @see `to_native_size()`
+    /// - @see `to_native_offset()`
+    /// - @see `to_arrow_length()`
+    /// - @see `sum_arrow_offsets()`
+    using arrow_length = std::int64_t;
```
Comment (DerThorsten) [inline include/sparrow/types/arrow_length.hpp:36]: if only internally it should be in a detail namespace

File: include/sparrow/types/arrow_length.hpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    /// Clarifies if a length value is supposed to be a size/length or an offset.
+    /// Offsets can be negative, sizes cannot. This is only important for runtime checks
```
Comment (DerThorsten) [inline include/sparrow/types/arrow_length.hpp:39]: in which context can offsets be negative?

File: include/sparrow/layout/run_end_encoded_layout/run_end_encoded_array.hpp
Comment (DerThorsten) [inline include/sparrow/layout/run_end_encoded_layout/run_end_encoded_array.hpp:1]: nitpicking: unrelated format changes should not be in this PR

File: test/test_arrow_length.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        CHECK_FALSE(is_valid_arrow_length(std::numeric_limits<std::int32_t>::min()));
+        CHECK(is_valid_arrow_length(-1, arrow_length_kind::offset));
+        CHECK(is_valid_arrow_length(-1024, arrow_length_kind::offset));
+        CHECK(is_valid_arrow_length(std::numeric_limits<std::int32_t>::min(), arrow_length_kind::offset));
+
+        CHECK(is_valid_arrow_length(std::size_t{0}));
+        CHECK(is_valid_arrow_length(std::size_t{1}));
+        CHECK(is_valid_arrow_length(std::ptrdiff_t{0}));
+        CHECK(is_valid_arrow_length(std::ptrdiff_t{1}));
+        CHECK(is_valid_arrow_length(std::ptrdiff_t{-1}, arrow_length_kind::offset));
```
Comment (DerThorsten) [inline test/test_arrow_length.cpp:49]: I believe there are no negative offsets in arrow

**Review by DerThorsten — CHANGES_REQUESTED** [review]

File: include/sparrow/config/config.hpp [hunk 4]
```diff
@@ -38,7 +38,37 @@
...
 {
-    return static_cast<bool>(COMPILING_WITH_APPLE_CLANG);
+    consteval bool is_apple_compiler()
+    {
+        return static_cast<bool>(COMPILING_WITH_APPLE_CLANG);
+    }
+
+    namespace config
+    {
+        inline constexpr bool enable_size_limit_runtime_check =
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/config/config.hpp:50]: this is for testing / debugging purpose, it is disabled by default in Release.

File: include/sparrow/types/arrow_length.hpp (hunk 2 above)
Comment (JohanMabille) [reply include/sparrow/types/arrow_length.hpp]: They cannot, I removed `arrow_length_kind`

**Review by DerThorsten — CHANGES_REQUESTED** [review]

File: include/sparrow/types/arrow_length.hpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /// @returns The sum of the provided offsets with `R` representation, whatever the offset types as long as
+    ///          they are integrals and can represent and arrow length.
+    ///          If `config::enable_size_limit_runtime_check == true` it will also check that each value and the resulting sum are
+    ///          valid arrow lengths and representable in the specified result `R` or throws otherwise.
+    /// @throws  @see `throw_if_invalid_size()` for details.
+    template<std::integral R = arrow_length>
+    inline constexpr
+    auto sum_arrow_offsets(std::integral auto... offsets) -> R
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/types/arrow_length.hpp:143]: We have some calls like [here](https://github.com/man-group/sparrow/blob/2baa1ed03f9cc8a0018e9e0f06307ad6ea4bda46/include/sparrow/buffer/buffer_adaptor.hpp#L481)

[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/types/arrow_length.hpp:143]: ? I dont understand  your comment at all...

my question is, why do we need a variadic function for that

[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/types/arrow_length.hpp:143]: Yeah forget about the example, it is unrelated. I'll replace it with a binary function.

File: include/sparrow/config/config.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Klaim) [reply include/sparrow/config/config.hpp:50]: > If a user is using 32 bit its, its in their responsibility to not overflow

We provide this so that users that use 64bit systems but want the data to be limited to 32bit can check that it is in fact the case and not discover the issues too late. And for testing purpose indeed. It is the same logic as allowing the user to enable assertions on their side.

File: include/sparrow/types/arrow_length.hpp (hunk 2 above)
Comment (Klaim) [reply include/sparrow/types/arrow_length.hpp]: Offsets in C++ standard can be negative, this is used to find a new index given a position or to compare positions and obtain a potentially negative (relative to the first element) offset. That's why `std::ptrdiff_t` is signed. 
Note that this is about the C++ standard side of things, not the Arrow side.
I didnt initially needed this distinction until I encountered lots of code in sparrow that was mixing `ptrdiff_t` and `size_t` to compute a position.

File: include/sparrow/types/arrow_length.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Klaim) [reply include/sparrow/types/arrow_length.hpp]: It's best to check the usage of these functions in [the previous PR](https://github.com/man-group/sparrow/pull/179/files#diff-6d1e508525bb97f11d4a493850e0fe5497c77c8269d4e28be88a254e069c3158L341) they are coming from to better understand why I ended up adding it.
Another example: https://github.com/man-group/sparrow/pull/179/files#diff-750d798de01fd1871b3e5e355d4c9b238d05a97a2e9a6975a40ca8b725b9bed5R157

You can find many examples in there.

While this function is not mandatory, it's utility is to:
- make sure that all the argument of such sum are in fact valid arrow length (which was not always the case in the previous pr, so that helped fixed that);
- when runtime checks are enabled, this is a check point to make sure the computed size is indeed in the required range (64bit or 32bit if the user want to constrain to it).

> But probably nowhere in the sparrow codebase we would use this, as this implies the length of the offsets (ie the number of variadic arguments) is known at compile time

In the usage from the previous PR you can see that we are always using 2 arguments with this functions, so it could be a 2 argument function, if you prefer. That wouldnt be specially superior to such version that takes as many argument as you need, but given the usage that wouldnt change a thing.

**Comment by DerThorsten** [comment]
I would suggest we add a free function to check sparrow arrays for 32bit compat.
`ensure_32_bit_compatibility(array)` which will raise if the array is to big for 32bit.
This function is always compiled, ie no config like `SPARROW_ENABLE_32BIT_SIZE_LIMIT/SPARROW_ENABLE_SIZE_CHECKS` is needed.
This function should just check the size of all buffers fits into 32bit.

This design has the benefit that we
 a) not need to continously check for 32bit violations at each and every considerable place.
 b) do not need an extra build where the size limits are enabled.


I believe its easier to just add this one function in a a separate PR, as not much (nothing?) of that PR would be left
